include/sys/trace.h

```diff
@@ -43,19 +43,23 @@
  */
 #undef DTRACE_PROBE1
 #define	DTRACE_PROBE1(name, t1, arg1) \
-	trace_zfs_##name((arg1))
+	if (arg1) \
+		trace_zfs_##name((arg1))
 
 #undef DTRACE_PROBE2
 #define	DTRACE_PROBE2(name, t1, arg1, t2, arg2) \
-	trace_zfs_##name((arg1), (arg2))
+	if (arg1 && arg2) \
```

Comment:
>  It might be possible to use the same conditional logic for printing custom fields on the arguments themselves in the dbuf__evict__one definition but I haven't tried if that works.

The following patch should allow us to keep tracepoints filtering capabilities while also being able to conditionally change the print format used in `TP_printk()`. On the other hand this will also increase the probe-effect since we will always have to `snprintf()` in `TP_fast_assign()` even for events that could potentially be filtered out.

```diff
diff --git a/include/sys/trace_dbuf.h b/include/sys/trace_dbuf.h
index edaa77f..f6e8b96 100644
--- a/include/sys/trace_dbuf.h
+++ b/include/sys/trace_dbuf.h
@@ -41,9 +41,22 @@
  *     zio_t *, ...);
  */
 
+#ifndef	TRACE_DBUF_MSG_MAX
+#define	TRACE_DBUF_MSG_MAX	255
+#endif
+
+#define	DBUF_TP_PRINTK_FMT                                              \
+	"dbuf { spa \"%s\" objset %llu object %llu level %llu "         \
+	"blkid %llu offset %llu size %llu state %llu holds %lld }"
+
+#define	DBUF_TP_PRINTK_ARGS					\
+	__get_str(os_spa), __entry->ds_object,                  \
+	__entry->db_object, __entry->db_level,                  \
+	__entry->db_blkid, __entry->db_offset,                  \
+	__entry->db_size, __entry->db_state, __entry->db_holds
+
 #define	DBUF_TP_STRUCT_ENTRY					\
-	__string(os_spa,					\
-	    spa_name(DB_DNODE(db)->dn_objset->os_spa))		\
+	__dynamic_array(char,	os_spa,	TRACE_DBUF_MSG_MAX)	\
 	__field(uint64_t,	ds_object)			\
 	__field(uint64_t,	db_object)			\
 	__field(uint64_t,	db_level)			\
@@ -52,31 +65,37 @@
 	__field(uint64_t,	db_size)			\
 	__field(uint64_t,	db_state)			\
 	__field(int64_t,	db_holds)			\
+	__dynamic_array(char,	msg,	TRACE_DBUF_MSG_MAX)
 
 #define	DBUF_TP_FAST_ASSIGN					\
-	__assign_str(os_spa,					\
-	    spa_name(DB_DNODE(db)->dn_objset->os_spa));		\
-								\
-	__entry->ds_object = db->db_objset->os_dsl_dataset ?	\
-	    db->db_objset->os_dsl_dataset->ds_object : 0;	\
-								\
-	__entry->db_object = db->db.db_object;			\
-	__entry->db_level  = db->db_level;			\
-	__entry->db_blkid  = db->db_blkid;			\
-	__entry->db_offset = db->db.db_offset;			\
-	__entry->db_size   = db->db.db_size;			\
-	__entry->db_state  = db->db_state;			\
-	__entry->db_holds  = refcount_count(&db->db_holds);
-
-#define	DBUF_TP_PRINTK_FMT						\
-	"dbuf { spa \"%s\" objset %llu object %llu level %llu "		\
-	"blkid %llu offset %llu size %llu state %llu holds %lld }"
+	if (db != NULL) {					\
+		__assign_str(os_spa,				\
+		    spa_name(DB_DNODE(db)->dn_objset->os_spa));	\
+		__entry->ds_object = db->db_objset->os_dsl_dataset ?	\
+		    db->db_objset->os_dsl_dataset->ds_object : 0;	\
+		__entry->db_object = db->db.db_object;		\
+		__entry->db_level  = db->db_level;		\
+		__entry->db_blkid  = db->db_blkid;		\
+		__entry->db_offset = db->db.db_offset;		\
+		__entry->db_size   = db->db.db_size;		\
+		__entry->db_state  = db->db_state;		\
+		__entry->db_holds  = refcount_count(&db->db_holds);	\
+		snprintf(__get_str(msg), TRACE_DBUF_MSG_MAX,	\
+		    DBUF_TP_PRINTK_FMT, DBUF_TP_PRINTK_ARGS);	\
+	} else {						\
+		__assign_str(os_spa, "NULL")			\
+		__entry->ds_object = 0;				\
+		__entry->db_object = 0;				\
+		__entry->db_level  = 0;				\
+		__entry->db_blkid  = 0;				\
+		__entry->db_offset = 0;				\
+		__entry->db_size   = 0;				\
+		__entry->db_state  = 0;				\
+		__entry->db_holds  = 0;				\
+		snprintf(__get_str(msg), TRACE_DBUF_MSG_MAX,	\
+		    "dbuf { NULL }");				\
+	}
 
-#define	DBUF_TP_PRINTK_ARGS					\
-	__get_str(os_spa), __entry->ds_object,			\
-	__entry->db_object, __entry->db_level,			\
-	__entry->db_blkid, __entry->db_offset,			\
-	__entry->db_size, __entry->db_state, __entry->db_holds
 
 /* BEGIN CSTYLED */
 DECLARE_EVENT_CLASS(zfs_dbuf_class,
@@ -84,7 +103,7 @@ DECLARE_EVENT_CLASS(zfs_dbuf_class,
 	TP_ARGS(db, zio),
 	TP_STRUCT__entry(DBUF_TP_STRUCT_ENTRY),
 	TP_fast_assign(DBUF_TP_FAST_ASSIGN),
-	TP_printk(DBUF_TP_PRINTK_FMT, DBUF_TP_PRINTK_ARGS)
+	TP_printk("%s", __get_str(msg))
 );
 /* END CSTYLED */
 
@@ -102,7 +121,7 @@ DECLARE_EVENT_CLASS(zfs_dbuf_evict_one_class,
 	TP_ARGS(db, mls),
 	TP_STRUCT__entry(DBUF_TP_STRUCT_ENTRY),
 	TP_fast_assign(DBUF_TP_FAST_ASSIGN),
-	TP_printk(DBUF_TP_PRINTK_FMT, DBUF_TP_PRINTK_ARGS)
+	TP_printk("%s", __get_str(msg))
 );
 /* END CSTYLED */
 
```

This results in:

```
root@linux:/usr/src/zfs# echo 0 > /sys/kernel/debug/tracing/trace
root@linux:/usr/src/zfs# echo 'os_spa == "NULL"' > /sys/kernel/debug/tracing/events/zfs/zfs_dbuf__evict__one/filter 
root@linux:/usr/src/zfs# echo 1 > /sys/kernel/debug/tracing/events/zfs/zfs_dbuf__evict__one/enable 
root@linux:/usr/src/zfs# cat /sys/kernel/debug/tracing/trace
# tracer: nop
#
# entries-in-buffer/entries-written: 5/5   #P:8
#
#                              _-----=> irqs-off
#                             / _----=> need-resched
#                            | / _---=> hardirq/softirq
#                            || / _--=> preempt-depth
#                            ||| /     delay
#           TASK-PID   CPU#  ||||    TIMESTAMP  FUNCTION
#              | |       |   ||||       |         |
      dbuf_evict-14559 [004] ....  8730.442979: zfs_dbuf__evict__one: dbuf { NULL }
      dbuf_evict-14559 [006] ....  8762.791772: zfs_dbuf__evict__one: dbuf { NULL }
      dbuf_evict-14559 [006] ....  8762.973285: zfs_dbuf__evict__one: dbuf { NULL }
      dbuf_evict-14559 [006] ....  8762.989567: zfs_dbuf__evict__one: dbuf { NULL }
      dbuf_evict-14559 [006] ....  8762.989959: zfs_dbuf__evict__one: dbuf { NULL }
```